src/main/java/org/opensearch/neuralsearch/search/collector/HybridTopFieldDocSortCollector.java

```diff
@@ -0,0 +1,403 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.search.collector;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.Locale;
+import java.util.ArrayList;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.log4j.Log4j2;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.Sort;
+import org.apache.lucene.search.FieldValueHitQueue;
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.TotalHits;
+import org.apache.lucene.search.LeafCollector;
+import org.apache.lucene.search.LeafFieldComparator;
+import org.apache.lucene.search.SortField;
+import org.apache.lucene.search.TopFieldDocs;
+import org.apache.lucene.search.Scorable;
+import org.apache.lucene.search.FieldComparator;
+import org.apache.lucene.search.CollectionTerminatedException;
+import org.apache.lucene.search.FieldDoc;
+import org.apache.lucene.util.PriorityQueue;
+import org.opensearch.neuralsearch.query.HybridQueryScorer;
+import org.opensearch.common.Nullable;
+import org.opensearch.neuralsearch.search.HitsThresholdChecker;
+import org.opensearch.neuralsearch.search.lucene.MultiLeafFieldComparator;
+
+/*
+ Collects the TopFieldDocs after executing hybrid query. Uses HybridQueryTopDocs as DTO to handle each sub query results.
+ The individual query results are sorted as per the sort criteria sent in the search request.
+ */
+@Log4j2
+public abstract class HybridTopFieldDocSortCollector implements Collector {
```

Comment:
I am failing to understand why we need this abstract class , Since Collector is already interface which has 3 functions 
1. getLeafCollector 2. scoreMode and 3. setWeight 

now Why dont' we have two class SingleFieldCollector and PagingField collector , directly implements collector infetrface and then a common helper class which implements leafCollector which will be intialized from above two classes to get leafCollector.

and when we are initalizing the class , intialize through interface not like new Class()

Reply:
It is the change in the design from TopFieldCollector from lucene. 
By your suggestion
A helper class will be an abstract class because collect method needs to be defined in the collector not in the helper class. That is what I have exactly done now.

I have created an abstract class which implements Collector. 
The collector method has 3 interface getLeafCollector, scoreMode and setWeight

getLeafCollector definition is different in case of PagingFieldColletor(sort, after) and in case of SimpleFieldCollector(sort, null) 

In the Leafcollector also the collect method implementation is different for both the cases. So there main collector class HybridTopFieldDocSortCollector will also be abstract and the leafCollector class HybridTopFieldDocLeafCollector class is also abstract.

Now the reason why HybridTopFieldDocLeafCollector cannot be a standalone class because it has lot of operations on the variables defined in the main collector class. Maintaining those variable states for every segment and shard is very difficult for every shard and segments.

Reply:
Can you check this class once https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java?